src/box/tuple_extract_key.cc

```diff
@@ -530,3 +530,51 @@ tuple_validate_key_parts(struct key_def *key_def, struct tuple *tuple)
 	}
 	return 0;
 }
+
+int
+tuple_validate_key_parts_raw(struct key_def *key_def, const char *tuple)
+{
+	assert(!key_def->is_multikey);
+	const uint32_t region_svp = region_used(&fiber()->gc);
+	field_map_builder builder;
+	int rc = tuple_field_map_create(tuple_format_runtime, tuple, false,
+					&builder);
+	if (rc != 0) {
+		region_truncate(&fiber()->gc, region_svp);
+		return -1;
+	}
+	uint32_t field_map_size = field_map_build_size(&builder);
+	char *buf = (char *)xregion_alloc(&fiber()->gc, field_map_size);
+	field_map_build(&builder, buf);
+	buf += field_map_size;
+	const uint32_t *field_map = (const uint32_t *)buf;
+
+	struct key_part *part = NULL;
+	const char *field = NULL;
+	for (uint32_t idx = 0; idx < key_def->part_count; idx++) {
+		part = &key_def->parts[idx];
+		field = tuple_field_raw_by_part(tuple_format_runtime, tuple,
+						field_map, part,
+						MULTIKEY_NONE);
+		if (field == NULL) {
+			if (key_part_is_nullable(part))
+				continue;
+			goto fail;
+		}
+		if (key_part_validate(part->type, field, idx,
+				      key_part_is_nullable(part)) != 0) {
+			diag_clear(diag_get());
+			goto fail;
+		}
+	}
+	region_truncate(&fiber()->gc, region_svp);
+	return 0;
+fail:
+	assert(part != NULL);
+	region_truncate(&fiber()->gc, region_svp);
+	diag_set(ClientError, ER_FIELD_TYPE,
+		 int2str(part->fieldno + TUPLE_INDEX_BASE),
```

Comment:
A field may have a path, which should be reported in the error (like it is in `tuple_validate_key_parts`). Please fix and add a test case.

Reply:
Done in a separate commit